### ISSUE TITLE
Union of dictionary and mapping not supported with pyright (but Pylance does)

### ISSUE THREAD

**Issue by jzazo** [microsoft/pyright#6941]
**Describe the bug**
When trying to merge dictionary and mapping pyright complains. If I add a type ignore comment, Pylance complaines that the comment is not needed.

**Code**
```python
from collections.abc import Mapping
from typing import cast

a = cast(Mapping[str, int], {"a": 1})
{"b": 2} | a  #  error: Operator "|" not supported for types "dict[str, int]" and "Mapping[str, int]" (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
I am running Pylance integrated in VSCode, and pyright in command line. The outputs differ.

Which version?
pyright: 1.1.345

**Comment by erictraut** [comment]
If you're seeing different behavior between pyright and pylance, that's because pylance is using an older version of pyright. The pylance team normally tries to minimize the difference between the two, but they're currently working on a complex change that's taking some time to complete. Once this change is in place, they'll be able to update to the latest version of pyright.

In the meantime, you could pin your version of pyright to an older version — the one that pylance is currently using.

**Comment by jzazo** [comment]
but the problem I reported occurs with latest pyright, which does not know how to merge dicts and mappings. Is this not a bug in pyright?

**Comment by erictraut** [comment]
No, pyright is doing the right thing here. The typeshed stub `builtins.pyi` defines the `dict` class as follows:

```python
class dict(MutableMapping[_KT, _VT]):
    ...
        @overload
        def __or__(self, __value: dict[_KT, _VT]) -> dict[_KT, _VT]: ...
        @overload
        def __or__(self, __value: dict[_T1, _T2]) -> dict[_KT | _T1, _VT | _T2]: ...
```

And the `Mapping` class doesn't support a `__ror__` method.

If you think that `dict` should accept a `Mapping` for this operation, please file an issue in the [typeshed project](https://github.com/python/typeshed). All of the major Python type checkers rely on typeshed for stdlib type definitions.

**Comment by jzazo** [comment]
Thanks, I will file an issue. However, Pylance does not complain as pyright does. I also pinned pyright to the same Pylance version to make sure, and it still complains.

**Comment by erictraut** [comment]
If you've pinned pyright to the same version that pylance is using and you're seeing different results, then you have configured pyright and pylance differently. To troubleshoot this issue, please refer to [this documentation](https://github.com/microsoft/pylance-release/blob/main/USING_WITH_PYRIGHT.md).

**Comment by jzazo** [comment]
I sorted the difference, I had wrong setting in `typeCheckingMode` in vscode. Thanks for the link.
